### DISCUSSION TITLE
AsyncSession commit not triggering before_flush event handler

### DISCUSSION THREAD

**Discussion by abhay-apella** [sqlalchemy/sqlalchemy#11714] [Usage Questions] [answered]
I have an application that I have migrated over from SQLAlchemy 1.4 over to SQLAlchemy 2.0. One of the behaviors we implemented was to add a history versioned row when we create new or updated rows. As part of this migration we moved over to async engine and for new code we use async sessions. The problem we notice is that when we invoke commit on an async session our `before_flush` listeners never get executed, but others such as `before_commit` or `after_commit` do. Manually setting `autoflush=True` on the session makers does not make any difference.

I declare my session maker like so:
```
sync_session_instance = sessionmaker(bind=engine.sync_engine, expire_on_commit=False)
async_session_instance = async_sessionmaker(
    bind=engine, expire_on_commit=False, sync_session_class=sync_session_instance
)
versioned_session(sync_session_instance)
```

I define event listeners like so:
```
def versioned_session(session: sessionmaker[Session]) -> None:
    @event.listens_for(session, "before_flush")
    def before_flush(session: Session, flush_context: object, instances: None) -> None:
        print(f"Before flush {session.new} {session.dirty} {session.deleted}", flush=True) # never gets here on an AsyncSession commit

    @event.listens_for(session, "after_commit")
    def after_commit(session: Session) -> None:
        print("After commit", flush=True) # DOES get here on an AsyncSession commit as well as before_commit
```
Abbreviated function code:
```
    async with async_session_instance() as session:
            session = session.sync_session
            greenlet_spawn(function_call, *args, **kwargs)
            ...

   def my_function_call(session: Session, my_arg1, my_arg2) -> None: 
           session.add(my_arg1)
           session.commit() # <- All event listeners get invoked

      
            session = session.sync_session
            greenlet_spawn(function_call, *args, **kwargs)
            ...

   async def my_async_function_call(session: Session, my_arg1, my_arg2) -> None: 
           async with async_session_instance() as session:
                 statement = statement.values([values_dict]).returning(MyArgType)  # type: ignore
                 returned_my_arg_type = (await session.scalars(statement)).one()
                 await session.commit() # <- before_flush does NOT get invoked, but before/after commit do get executed.
```

I have tried stepping through the implementation but I do not see why `before_flush` would not get invoked.  Flush is being invoked, since I can query the same objects in a different session and get the updated objects.
Is there something I am doing wrong? Is there bug or something else going on behind the scenes? What could I do to fix this?

**Comment by CaselIT** [comment]
Hi,

A few notes on your example:
```py
sync_session_instance = sessionmaker(bind=engine.sync_engine, expire_on_commit=False)
async_session_instance = async_sessionmaker(
    bind=engine, expire_on_commit=False, sync_session_class=sync_session_instance
)
```

Why are you defining both an async_sessionmaker and a normal sessionmaker? there should be no need to define both.

```py
async with async_session_instance() as session:
            session = session.sync_session
            greenlet_spawn(function_call, *args, **kwargs)
```

don't do that. Use the documented run_sync: https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncSession.run_sync

Regarding your issue it's may be that the use of two session makers is causing issues with the event. Can you try with the above changes to see if it keep happening?

**Reply by abhay-apella** [reply]
I was following the recommendation in one of the other topics ([this](https://github.com/sqlalchemy/sqlalchemy/discussions/6594#discussioncomment-836437) and [this](https://github.com/sqlalchemy/sqlalchemy/discussions/9403)). In order to add events on the sync session in different parts of my app I needed to expose I had to use a sync session maker reference. I did try different combinations there without the sync session maker but it would always error out since we cannot listen on async sessions.  One difference with sqlalchemy 2 we started to implement was a transition to statements instead of through session add/commit. So we do
```
async with async_session_instance() as session:
                 statement = statement.values([values_dict]).returning(MyArgType)  # type: ignore
                 returned_my_arg_type = (await session.scalars(statement)).one()
                 await session.commit() # <- before_flush does NOT get invoked, but before/after commit do get executed.
```
where the "flush" occurs as part of the `session.scalars` call, and the commit does not have any additional state to be flushed to the db. Looking at: https://docs.sqlalchemy.org/en/20/core/event.html#event-reference, it looks like I need to also add an event listener on `do_orm_execute`. On debugging I know the event listener on `do_orm_execute` is being invoked so I could add a check for if its an insert/update/delete then invoke my versioning call, but I am unable to grab the history mapper class that we attach to the regular object mapper. That being said should `session.execute` or `session.scalars` invoke `before_flush` as well?

**Reply by CaselIT** [reply]
> . I did try different combinations there without the sync session maker but it would always error out since we cannot listen on async sessions

You would register the events on a normal Session. 

Regarding execution, session will run a flush before any execute call (scalars calls execute) if needed. Off the top of my head I'm not sure of the before_flush event happens if there is nothing to flush. @zzzeek this doesn't seem mentioned in the event docs, maybe a line in before/after flush may me useful?

**Reply by abhay-apella** [reply]
Got it I tried:
```
@event.listens_for(Session, "before_flush")
    def before_flush(session: Session, flush_context: object, instances: None) -> None:
        print(f"Before flush {session.new} {session.dirty} {session.deleted}", flush=True) 
```
and still no luck for session.execute/scalars, but it does execute for 1.4 way ie. `session.add(some_obj)` and `session.commit()` :(. Stepping through the function it looks like the flush condition is checking if there is an object added to the session before flushing but that is not what I am doing with:
```
statement = statement.values([values_dict]).returning(MyArgType)
```
since I am just adding the insert statement, which is why `do_orm_execute` does execute. Should the solution be just to transition over to `do_orm_execute` or should I be adding the object to the session as well as the insert statement?

**Reply by CaselIT** [reply, marked as answer]
so looking at the code the flush event happen only if there is something to flush. executing insert statements directly will not cause flush to happen since it does not result in nothing to be flushed.

> should I be adding the object to the session as well as the insert statement?

that would add object two, so no. `do_orm_execute ` gets triggered when you execute statement directly though.

I'm not sure they you are changing from using the session directly to the insert pattern. Using the session directly is still perfectly supported and it's not legacy or discouraged in any way, so if you already had a working application using that pattern feel free to keep it that way. What I mean is don't feel obligated to use the insert pattern unless you want to

**Reply by abhay-apella** [reply]
That's great to know! We were following: https://docs.sqlalchemy.org/en/20/core/dml.html which seems to recommend updating to a new convention. Though when migrating to SqlAlchemy 2 some parts of our code are better served with the new statement pattern e.g. batch inserts with `session.add` didn't seem to work well from 1.4 to 2.0 and instead caused attempting to insert all values as a single row like attempting to inserting 5 objects caused `session.commit` to run 1 insert with all 5 objects values as 1 object. So it seems we need probably modify our versioning in order to work for 2.0. But for all other cases we will just keep as is to remove unnecessary work.

**Reply by CaselIT** [reply]
Note that that's the core part of sqlalchemy, that uses tables directly. If you use that you need to use insert and similar stamens direct, since there is no session there, since the core does not expose an orm.

Regarding the inset on multiple objects, in  v2 the session can use `insert ... Values (),(),...` to inset many rows at the same time, greatly improving performance

**Reply by abhay-apella** [reply]
Is that possible? When I try do something like:
```
for obj in my_obj_list:
    session.add(obj)

session.commit()
```

I get errors. The only documentation I seem to find  on batch inserts seems to be on the link above. Is there any docs I could read how to do orm batch insert or upserts?

**Reply by CaselIT** [reply]
> Is that possible

sure, there's also add_all that accepts a list of objects.

> I get errors

that's a bit vague. assuming `my_obj_list` includes a list of instances no error should happen. Unless you try to insert duplicate keys or similar constraint violations

**Reply by abhay-apella** [reply]
That works with an async session! We did migrate a bunch of adds in our old sync sessions that weren't working, even with add_all or merge. The old error used to be that if I am attempting to add 3 distinct objects of the same type each with 3 attributes it would only do a single `insert into my_table (col1, col2, col3) with values (o1v1, o1v2, o1v3, o1v1, o2v2, o2v3, o3v1, o3v2, o3v3)` instead of `(o1v1, o1v2, o1v3), (o1v1, o2v2, o2v3), (o3v1, o3v2, o3v3)`. I can revert the operations over to session level again!

**Reply by CaselIT** [reply]
Strange, but if you have fixed it now then great!
